src/app/reporting/Read-DataModel.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include <app/reporting/Read-DataModel.h>
+
+#include <app/AppConfig.h>
+#include <app/data-model-interface/DataModel.h>
+#include <app/util/MatterCallbacks.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+namespace DataModelImpl {
+namespace {
+
+bool IsOutOfSpaceError(CHIP_ERROR err)
+{
+    return (err == CHIP_ERROR_NO_MEMORY || err == CHIP_ERROR_BUFFER_TOO_SMALL);
+}
+
+} // namespace
+
+CHIP_ERROR RetrieveClusterData(InteractionModel::DataModel * dataModel, const Access::SubjectDescriptor & subjectDescriptor,
+                               bool isFabricFiltered, AttributeReportIBs::Builder & reportBuilder,
+                               const ConcreteReadAttributePath & path, AttributeEncodeState * encoderState)
+{
+#if !CHIP_CONFIG_USE_EMBER_DATA_MODEL
+    ChipLogDetail(DataManagement, "<RE:Run> Cluster %" PRIx32 ", Attribute %" PRIx32 " is dirty", path.mClusterId,
+                  path.mAttributeId);
+    DataModelCallbacks::GetInstance()->AttributeOperation(DataModelCallbacks::OperationType::Read,
+                                                          DataModelCallbacks::OperationOrder::Pre, path);
+#endif // !CHIP_CONFIG_USE_EMBER_DATA_MODEL
+
+    InteractionModel::ReadAttributeRequest readRequest;
+
+    if (isFabricFiltered)
+    {
+        readRequest.readFlags.Set(InteractionModel::ReadFlags::kFabricFiltered);
+    }
+    readRequest.subjectDescriptor = subjectDescriptor;
+    readRequest.path              = path;
+
+    DataVersion version = 0;
+    if (std::optional<InteractionModel::ClusterInfo> clusterInfo = dataModel->GetClusterInfo(path); clusterInfo.has_value())
+    {
+        version = clusterInfo->dataVersion;
+    }
+    else
+    {
+        ChipLogError(DataManagement, "Read request on unknown cluster - no data version available");
+    }
+
+    TLV::TLVWriter checkpoint;
+    reportBuilder.Checkpoint(checkpoint);
+
+    AttributeValueEncoder attributeValueEncoder(reportBuilder, subjectDescriptor, path, version, isFabricFiltered, encoderState);
+    CHIP_ERROR err = dataModel->ReadAttribute(readRequest, attributeValueEncoder);
+
+    if (err == CHIP_NO_ERROR)
+    {
+#if !CHIP_CONFIG_USE_EMBER_DATA_MODEL
+        DataModelCallbacks::GetInstance()->AttributeOperation(DataModelCallbacks::OperationType::Read,
```

Comment:
I know this preserves the old logic for when this callback is called, but that logic seems off to me....  Separate PR, though.

Reply:
The problem is exactly that the post read is not called on failure, so you can get pre-read a few times, then a post-read, and how do you know what that means?

More generally: what is the intent of these callbacks?  And does calling this only on success serve that intent?

Reply:
Added TODO. We can clean it up later (and probably much easier given the current placement).